File: lua/ht/conf/external_tool/formatters.lua
local formatter = require("ht.core.external_tool").formatter

---@type ht.external_tool.Formatter[]
local formatters = {}

formatters[#formatters + 1] = formatter.new(
  "c,cpp,cs,java,cuda,proto",
  "clang-format",
  {
    args = { "-style=file" },
    stdin = true,
  },
  {
    name = "clang-format",
  }
)

-- TODO(hawtian): stylua condition,
--       condition = function(utils)
--         return utils.root_has_file { "stylua.toml", ".stylua.toml" }
--       end,
formatters[#formatters + 1] = formatter.new("lua", "stylua", {
  args = { "--search-parent-directories", "-" },
  stdin = true,
}, {
  name = "stylua",
})

formatters[#formatters + 1] = formatter.new("rust", "rustfmt", {
  args = { "--emit", "stdout" },
  stdin = true,
}, {
  name = "rustfmt",
  mason = false,
})

formatters[#formatters + 1] = formatter.new(
  "javascript,javascriptreact,typescript,typescriptreact,vue,css,scss,less,html,json,jsonc,yaml,markdown,markdown.mdx,graphql,handlebars",
  "prettier",
  {
    args = { "--stdin-filepath", "--prose-wrap", "never" },
    stdin = true,
    fname = true,
  },
  {
    name = "prettier",
  }
)

formatters[#formatters + 1] = formatter.new("python", "black", {
  args = { "--quiet", "-" },
  stdin = true,
  find = "pyproject.toml",
}, {
  name = "black",
})

formatters[#formatters + 1] = formatter.new("dart", "dart", {
  args = { "format" },
  stdin = true,
}, {
  name = "dart",
  mason = false,
})

formatters[#formatters+1] = formatter.new("cmake", "gersemi", {
  args = { "-" },
  stdin = true,
}, {
  name = "gersemi"
})

return formatters

File: src/conf/external_tools/formatters.ts
import { Formatter } from "@core/model";

/**
 * Variables will be replaced.
 *
 * $FILENAME
 * $DIRNAME
 */
const ClangFormat = new Formatter({
  name: "clang-format",
  exe: false,
  ft: ["c", "cpp", "cs", "java", "cuda", "proto"],
  opts: {
    args: ["-assume-filename", "$FILENAME"],
    stdin: true,
  },
});

const Stylua = new Formatter({
  name: "stylua",
  exe: false,
  ft: ["lua"],
  opts: {
    args: ["--search-parent-directories", "--stdin-filepath", "$FILENAME", "-"],
    stdin: true,
  },
});

const RustFmt = new Formatter({
  name: "rustfmt",
  exe: false,
  ft: ["rust"],
  opts: {
    args: ["--emit=stdout"],
    stdin: true,
  },
});

const Prettier = new Formatter({
  name: "prettier",
  exe: false,
  ft: [
    "javascript",
    "javascriptreact",
    "typescript",
    "typescriptreact",
    "vue",
    "css",
    "scss",
    "less",
    "html",
    "json",
    "jsonc",
    "yaml",
    "markdown",
    "markdown.mdx",
    "graphql",
    "handlebars",
  ],
  opts: {
    args: ["--stdin-filepath", "$FILENAME"],
    stdin: true,
  },
});

const Black = new Formatter({
  name: "black",
  exe: false,
  ft: ["python"],
  opts: {
    args: ["--stdin-filename", "$FILENAME", "--quiet", "-"],
    stdin: true,
  },
});

const DartFmt = new Formatter({
  name: "dartfmt",
  exe: {
    command: ["dart"],
  },
  ft: ["dart"],
  opts: {
    args: ["format"],
    stdin: true,
  },
});

const Gersemi = new Formatter({
  name: "gersemi",
  exe: false,
  ft: ["cmake"],
  opts: {
    args: ["-"],
    stdin: true,
  },
});

const Alejandra = new Formatter({
  name: "alejandra",
  exe: false,
  ft: ["nix"],
});

const SwiftFormat = new Formatter({
  name: "swiftformat",
  exe: {
    command: ["/opt/homebrew/bin/swift-format"],
  },
  ft: ["swift"],
  opts: {
    args: ["$FILENAME", "--in-place"],
    stdin: false,
  },
});

export default [
  ClangFormat,
  Stylua,
  RustFmt,
  Prettier,
  Black,
  DartFmt,
  Gersemi,
  Alejandra,
  SwiftFormat,
];
